Better presentation of failure

bittrance opened the pull request:
- UPDATE_ROLLBACK_COMPLETE is hardly a good state; treat it as a failure
- rxcfclient treats bad states as Rx errors
- present checked errors by their message
- exit with 1 on errors

Fixes #6.
Fixes #32.

Commit:
Better presentation of failure.

- UPDATE_ROLLBACK_COMPLETE is hardly a good state; treat it as a failure
- rxcfclient treats bad states as Rx errors
- present checked errors by their message
- exit with 1 on errors

Fixes #6, #32.

grddev commented on lib/cuffsert/presenters.rb:
```diff
@@ -90,7 +101,7 @@ def on_stack_event(event)
       @renderer.event(event, resource)
       @renderer.clear
-      @resources.each { |resource| @renderer.resource(resource) }
+      @resources.each { |r| @renderer.resource(r) }
```
This seems accidental

bittrance replied:
Fixed.

grddev commented on lib/cuffsert/rxcfclient.rb:
```diff
@@ -94,8 +99,15 @@ def record_start_time
+      return false unless event[:physical_resource_id] == stack_id
+      case CuffSert.state_category(event[:resource_status])
+      when :bad
+        raise RxCFError, "Stack #{event.logical_resource_id} finished in state #{event.resource_status}"
```
Not sure how this all fits together, but won't raising here mean that we risk not processing additional events?

bittrance replied:
That's true, good catch. Concretely, cleanup-events will show up after the stack event. Not sure if you can get cleanup-events after a failure, but let's not take that chance.

bittrance replied:
Fixed.

Commit:
Ensure we do not drop events when the stack update fails.

As pointed out by grddev we need to process all events before raising bad stack
states.

Commit:
fixup! Better presentation of failure.

grddev commented on lib/cuffsert/presenters.rb:
```diff
@@ -76,6 +77,16 @@ def on_event(event)
+      else
+        super(err)
+      end
+      exit(1)
```
In general I don't like calling exit outside "the main function" as it means that the feature composes very badly (and makes it somewhat more difficult to write tests), but maybe it works as intended in this case.

bittrance replied:
I agree. However, Cuffsert uses the "official" RxRuby which has no "await" interface, so the alternative would be to use a mutex or some similar construct to block in `#main` until the presenter is finished. Since there is no indication that RxRuby is maintained, this would have to be a local solution. My experience with `Mutex`, `Queue` and `ConditionalVariable` is somewhat mixed, so in the end I opted for this ugliness.

grddev commented on lib/cuffsert/rxcfclient.rb:
```diff
@@ -57,8 +58,12 @@ def update_stack(stack_id, change_set_id)
+            observer.on_next(event)
+          end
+        rescue => e
+          observer.on_error(e)
```
Given that this calls `exit`, this will raise an error that means that `on_completed` will not be called. If that is intentional, I think it would be nicer to stick the on_completed call in an `else` block (or inside the `begin` block depending on the whether errors in the call to `on_completed` can be meaningfully handled using `on_error`.

bittrance replied:
Ah, good catch. The Rx contract is to call `on_error` or `on_completed` but not both. RxRuby has some guards that will ensure that `on_completed` will be a no-op in the error case, but you are absolute right that the code should not depend on that. Fixed.

grddev reviewed (approved):
Looks good overall. It seems a bit iffy to me to use `exit` in (non-top-level) error handling, but it can be changed later if needed.
